---
title: "Seurat Pipeline Analysis"
output: html_notebook
---

Script for analysing LAP_GEX 10x Chromium Single Cell Multiome ATAC + Gene Expression assay of a 56 year old Female. 

The 10x web summary file after running the Cell Ranger arc pipeline on Sample D can be found [here](/media/Home_Raid1_Voyager/dinh/OMICS_dinh/10XMULTI/LAPMAP20220202D/outs/web_summary.html).

## Setting up Workspace

Importing all relevant packages
```{r}
library(Seurat)
library(dplyr)
library(Matrix)
```

setting paths and count matrix 
```{r}
dir1 <- "/media/NAS2/Users/OMIC_ProcessedData/LAPMAP_ProcessedData/10X_snMulti-seqV1/"
dir2 <- paste0("/media/NAS3/elizabeth_NAS3/Analysis_Folders/LAPMAP_10XMulti/LAPMAP20220202_10XR_matrices/")
```

setting directory to dir2 (contains count matrix)
```{r}
setwd(dir2)
save(LAP_GEX_20220202D, file = "/media/NAS3/tanvi_NAS3/LAP_GEX_20220202D_10XMULTI.rda")
```
```{r}
###LAPMAP_20220202-10XMulti
sample="LAPMAP_20220202D"
sample <- Read10X(paste0(dir1, sample, "/outs/filtered_feature_bc_matrix/"))
sample_GEX=sample$`Gene Expression`

dim(sample_GEX) #36601  9087
head(colnames(sample_GEX))
colnames(sample_GEX) <- paste("20220202D", colnames(sample_GEX), sep="_")
LAP_GEX_20220202D=sample_GEX
# sample_GEX="LAPMAP_20211202B_10X-R"

saveRDS(LAP-GEX_20220202D, file = paste0(dir2, sample, "_UMI_counts.rds"))
```

reading count matrix
```{r}
###LAPMAP_20220202-10XMulti
sample="LAPMAP_20220202D"
sample <- Read10X("/media/NAS3/tanvi_NAS3/")
sample_GEX=sample$`Gene Expression`
```
Checking dimensions of count matrix. Checking head and colnames of count matrix
```{r}
dim(sample_GEX) #36601  9087
head(colnames(sample_GEX))
colnames(sample_GEX) <- paste("20220202D", colnames(sample_GEX), sep="_")
LAP_GEX_20220202D=sample_GEX
# sample_GEX="LAPMAP_20211202B_10X-R"
```
Creating Seurat Object
```{r}
saveRDS(LAP-GEX_20220202D, file = paste0(dir2, sample, "_UMI_counts.rds"))
LAP_GEX_20220202D <- CreateSeuratObject(counts = sample_GEX, project = "LAP_GEX_20220202D_Seurat", min.cells = 3, min.features = 200)
```

## QC Filtering
to filter out empty and doublet cells, high % mitochondrial genes
```{r}
LAP_GEX_20220202D[["percent.mt"]] <- PercentageFeatureSet(LAP_GEX_20220202D, pattern = "^MT-")
head(LAP_GEX_20220202D@meta.data, 5)
```
```{r}
levels(Idents(LAP_GEX_20220202D))
table(LAP_GEX_20220202D$orig.ident)
Idents(LAP_GEX_20220202D)<-"orig.ident"
levels(Idents(LAP_GEX_20220202D))
```

Visualising via violin plots to visualise the thresholds for filtering out high% mitochondrial gene cells. 
The Feature Scatter Plot should show us if genes are not deeply seqeunced (dots on the top left corner), which is expected from out 10x summary report, and cells on the bottom right corner (repetitive/doublets?)
nFeature_RNA - number of genes detected in every cell
nCount_RNA - total number of molecules detected within a cell
percent.mt -percentage of reads that map to the mitochondrial gene
```{r}
VlnPlot(LAP_GEX_20220202D, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), ncol=3, pt.size=0)

```
```{r}
VlnPlot(LAP_GEX_20220202D, features ="nFeature_RNA", ncol=3, pt.size=0, y.max = 7500)
```
```{r}
metadata <- LAP_GEX_20220202D@meta.data
summary(metadata$nFeature_RNA)
```
```{r}
summary(metadata$nCount_RNA)
```

```{r}
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

FeatureScatter(LAP_GEX_20220202D, feature1 = "nCount_RNA", feature2 = "percent.mt")
#
```
```{r}
FeatureScatter(LAP_GEX_20220202D, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```

```{r}
# 200-5000 for nFeature_RNA?
# Vignette Seurat values are 200-2500
nFeature_RNA_lower <- 200
nFeature_RNA_upper <- 7500
percent_mt_upper <- 5

LAP_GEX_20220202D <- subset(LAP_GEX_20220202D, subset = nFeature_RNA > nFeature_RNA_lower & nFeature_RNA < nFeature_RNA_upper & percent.mt < percent_mt_upper)
```

## Normalization
accounts for differences in sequencing depths
```{r}
LAP_GEX_20220202D <- NormalizeData(LAP_GEX_20220202D)
```

## Identification of highly variable genes
```{r}
LAP_GEX_20220202D <- FindVariableFeatures(LAP_GEX_20220202D, selection.method = "vst", nfeatures = 2000)
```

```{r}
# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(LAP_GEX_20220202D), 10)
top10
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(LAP_GEX_20220202D)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2
```
```{r}
LabelPoints(plot = plot1, points = top10, repel = TRUE)
```

## Scaling Data

```{r}
all.genes <- rownames(LAP_GEX_20220202D)
LAP_GEX_20220202D <- ScaleData(LAP_GEX_20220202D, features = all.genes)
```

## Linear Reduction Dimension
```{r}
LAP_GEX_20220202D <- RunPCA(LAP_GEX_20220202D, features = VariableFeatures(object = LAP_GEX_20220202D))
```

```{r}
# Examine and visualize PCA results a few different ways
print(LAP_GEX_20220202D[["pca"]], dims = 1:5, nfeatures = 5)
```
look out for abnormal genes
```{r}
VizDimLoadings(LAP_GEX_20220202D, dims = 1:2, reduction = "pca")
```

```{r}
DimPlot(LAP_GEX_20220202D, reduction = "pca")
```

```{r}
DimHeatmap(LAP_GEX_20220202D, dims = 1, cells = 500, balanced = TRUE)
```

```{r}
DimHeatmap(LAP_GEX_20220202D, dims = 1:15, cells = 500, balanced = TRUE)
```

## Determining Dimensionality of Data
```{r}
# NOTE: This process can take a long time for big datasets, comment out for expediency. More
# approximate techniques such as those implemented in ElbowPlot() can be used to reduce
# computation time
pbmc <- JackStraw(pbmc, num.replicate = 100)
pbmc <- ScoreJackStraw(pbmc, dims = 1:20)
```

```{r}
JackStrawPlot(pbmc, dims = 1:15)
```
plot it out further to 50 PCs to see where it flattens out
```{r}
ElbowPlot(LAP_GEX_20220202D)

```

## Cell Clustering
Check out the parameters chosen 
```{r}
LAP_GEX_20220202D <- FindNeighbors(LAP_GEX_20220202D, dims = 1:10)
LAP_GEX_20220202D <- FindClusters(LAP_GEX_20220202D, resolution = c(0.5,0.75,1.0,1.2,1.4,1.6,1.8,2.0))
```
```{r}
LAP_GEX_20220202D[[]]
```
 
```{r}
# Look at cluster IDs of the first 5 cells
head(Idents(LAP_GEX_20220202D), 5)
```

## Non-Linear Dimensional Reduction
```{r}
# If you haven't installed UMAP, you can do so via reticulate::py_install(packages =
# 'umap-learn')
LAP_GEX_20220202D <- RunUMAP(LAP_GEX_20220202D, dims = 1:10)
```

```{r}
# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
Idents(LAP_GEX_20220202D) <- 'RNA_snn_res.0.5'
DimPlot(LAP_GEX_20220202D, reduction = "umap",label=TRUE)
```
```{r}
Idents(LAP_GEX_20220202D) <- 'RNA_snn_res.1'
DimPlot(LAP_GEX_20220202D, reduction = "umap",label=TRUE)
```
```{r}
Idents(LAP_GEX_20220202D) <- 'RNA_snn_res.1.6'
DimPlot(LAP_GEX_20220202D, reduction = "umap",label=TRUE)
```

## Differentially Expressed gene analysis
FindAllMarkers()
logfc.threshold
wilcox test -> The Wilcoxon test checks whether the mean values of two dependent groups differ significantly from each other. The Wilcoxon test is a non-parametric test and is therefore subject to considerably fewer assumptions than its parametric counterpart, the t-test for dependent samples. Therefore, as soon as the boundary conditions for the t-test for dependent samples are no longer fulfilled, the Wilcoxon test is used.
```{r}
Idents(LAP_GEX_20220202D) <- 'RNA_snn_res.1'
LAP_GEX_20220202D.markers <- FindAllMarkers(LAP_GEX_20220202D, assay = 'RNA',features = NULL, logfc.threshold = 0.25, test.use="wilcox",slot="data",min.pct=0.1,min.diff.pct=-Inf,node = NULL, verbose = TRUE, only.pos = FALSE, max.cells.per.ident = Inf,random.seed = 1, latent.vars = NULL, min.cells.feature = 3, min.cells.group = 3, mean.fxn = NULL,fc.name = NULL, base = 2, return.thresh = 0.01,densify = FALSE)
```
```{r}
saveRDS(LAP-GEX_20220202D, file = paste0(dir2, sample, "_findAllMarkers.rds"))
```
```{r}
LAP_GEX_20220202D.markers <- 
```

```{r}
LAP_GEX_20220202D.markers_sorted <- LAP_GEX_20220202D.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
LAP_GEX_20220202D.markers_sorted <- unique(LAP_GEX_20220202D.markers_sorted$gene)
DotPlot(object = LAP_GEX_20220202D, features = LAP_GEX_20220202D.markers_sorted) + RotatedAxis()
```
```{r}
Idents(LAP_GEX_20220202D) <- 'RNA_snn_res.1'
markergenes <- read.csv("/media/Scratch_SSD_Voyager/Tanvi/R/lungmarkers.csv")
#LAP_GEX_20220202D.markers_sorted <- LAP_GEX_20220202D.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
fibroblast_markergenes <- unique(markergenes[,1])
DotPlot(object = LAP_GEX_20220202D, features = fibroblast_markergenes) + RotatedAxis() + theme(axis.text = element_text(size = 7))  
```
#Labelling Cell types using CellRef table
```{r}
#intersect(LAP_GEX_20220202D.markers[LAP_GEX_20220202D.markers$cluster==5,]$gene,
lung_markers_reference <- read.csv("/media/Scratch_SSD_Voyager/Tanvi/R/CellRef.csv", header=TRUE)
library(ggplot2)
cell_type <- 'AEC'
#Validating via FindAllMarkers() table
marker_genes_for_celltype <- lung_markers_reference[lung_markers_reference$Cell.type==cell_type,]$Marker
LAP_GEX_20220202D.markers[LAP_GEX_20220202D.markers$gene %in% marker_genes_for_celltype,]

#Feature Plot
Idents(LAP_GEX_20220202D) <- 'RNA_snn_res.1'
# Feature plot - visualize feature expression in low-dimensional space
FeaturePlot(LAP_GEX_20220202D, features = marker_genes_for_celltype)

#DotPlot
DotPlot(object = LAP_GEX_20220202D, features = marker_genes_for_celltype) + RotatedAxis() + theme(axis.text = element_text(size = 7)) 
#Vln Plot

```
```{r}
Idents(LAP_GEX_20220202D) <- 'RNA_snn_res.1'
# Feature plot - visualize feature expression in low-dimensional space
FeaturePlot(LAP_GEX_20220202D, features = c('AGER','RTKN2','SEMA3B'))

```
```{r}
DotPlot(object = LAP_GEX_20220202D, features = c('AGER','RTKN2','SEMA3B')) + RotatedAxis() + theme(axis.text = element_text(size = 7))  
```

```{r}
LAP_GEX_20220202D.markers_sorted[duplicated(LAP_GEX_20220202D.markers_sorted)]
```

```{r}
write.csv(LAP_GEX_20220202D.markers, "/media/Scratch_SSD_Voyager/Tanvi/LAP_GEX_20220202DMarkers.csv", row.names=TRUE)
```

use violin or dot plot too
```{r}
LAP_GEX_20220202D.markers
write.csv(LAP_GEX_20220202D.markers, "/media/Scratch_SSD_Voyager/Tanvi/R/LAP_markers.csv")
saveRDS(LAP_GEX_20220202D.markers ,file="LAP_markers.RData")
```

```{r}
lung_markers <- read.table("/media/Scratch_SSD_Voyager/Tanvi/LungMarkers.csv")
#FeaturePlot(object = LAP_GEX_20220202D, features = c('ACTA2','DES','LGR6','NTRK3','ITGA7','COL2A1','HAPLN1','TCF21','WNT21','MFAP5','SCARA5'), cols = c("grey", "red"), raster=FALSE, reduction = "umap")
```
```{r}
DotPlot(
  LAP_GEX_20220202D,
  assay = "RNA",
  features = LAP_GEX_20220202D_dotfeatures,
  cols = c("lightgrey", "blue"),
  col.min = -2.5,
  col.max = 2.5,
  dot.min = 0,
  dot.scale = 6,
  group.by = NULL,
  split.by = NULL,
  cluster.idents = FALSE,
  scale = TRUE,
  scale.by = "radius",
  scale.min = NA,
  scale.max = NA
)
```
```{r}
VlnPlot(
  LAP_GEX_20220202D,
  features = lung_markers[1,1:nrow(lung_markers)],
  cols = NULL,
  pt.size = NULL,
  idents = 'RNA_snn_res.1',
  sort = FALSE,
  assay = 'RNA',
  group.by = NULL,
  split.by = NULL,
  adjust = 1,
  y.max = NULL,
  same.y.lims = FALSE,
  log = FALSE,
  ncol = NULL,
  slot = "data",
  split.plot = FALSE,
  stack = FALSE,
  combine = TRUE,
  fill.by = "feature",
  flip = FALSE,
  add.noise = TRUE,
  raster = NULL
)
```

pick middle level (resolution 1) use cell cards for references. Keep track of markers that show evidence for each cluster
```{r}
LAP_GEX_20220202D_labelled_res0.5 <-RenameIdents(LAP_GEX_20220202D, `0`   = "unknown",
                                                              `1`   = "unknown",
                                                              `2`   = "unknown",
                                                              `3`   = "unknown",
                                                              `4`   = "unknown",
                                                              `5`   = "unknown",
                                                              `6`   = "AT2",
                                                              `7`   = "AT1",
                                                              `8`   = "Ciliated",
                                                              `9`   = "Myoepithelial and muscle",
                                                              `10`   = "unknown",
                                                              `11`   = "Secretory mucosal ",
                                                              `12`   = "Basal",
                                                              `13`   = "Ciliated")
```

```{r}
saveRDS(LAP_GEX_20220202D, file = "../output/pbmc_tutorial.rds")
```

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file). 

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.
